File: python/151-160/Find Minimum in Rotated Sorted Array.py
class Solution:
    # @param num, a list of integer
    # @return an integer
    def findMin(self, num):
        size=len(num)
        if size==0:
            return 0
        Min=num[0]
        low,high=0,size-1
        while low<=high:
            mid = low+high//2
            print low,high,mid
            if num[mid]<Min:
                Min=num[mid]
                if num[mid-1]<num[mid]:
                    high=mid-1
                else:
                    break
            else: 
                low=mid+1
        return Min
so=Solution()
data=[1,2,3]
print so.findMin(data)

File: Cpp/151-160/Find Minimum in Rotated Sorted Array.cpp

/*
    binary search method 
    o(lgn) time 
    o(1) space 
    run time : 4ms
    error :1   in line : while(beg<end)
*/
class Solution {
public:
    int findMin(vector<int>& nums) {
        int beg = 0,end = nums.size()-1;
        while(beg<end){
          int mid = beg+(end-beg)/2;
          if(nums[mid]>nums[end])
            beg = mid+1;
          else if(nums[mid]<nums[end])
            end = mid;
          else end--;
        }
        return nums[beg];
    }
};
